### PR TITLE
Implement iterative DMatrix.

### PR DESCRIPTION
**Pull request by trivialfis**
Extracted from #5783 .

In the big PR, device DMatrix is removed and replaced by iterative DMatrix.  This PR hasn't removed it yet so I can extract the related parts without breaking XGBoost.

### CODE REVIEW

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -221,6 +155,189 @@ XGB_DLL int XGDMatrixCreateFromDT(void** data,
                                   bst_ulong ncol,
                                   DMatrixHandle* out,
                                   int nthread);
+
+/*
+ * ========================== Begin data callback APIs =========================
+ *
+ * Short notes for data callback
+ *
+ * There are 2 sets of data cakkbacks for DMatrix.  The first one is currently exclusively
```
Comment (RAMitchell) [inline include/xgboost/c_api.h:164]: "cakkbacks" :)

This improved documentation is really nice.

File: include/xgboost/data.h [hunk 2]
```diff
@@ -502,7 +502,33 @@ class DMatrix {
...
+   * \param next    Callback for next
+   * \param missing Value that should be treated as missing.
+   * \param nthread number of threads used for initialization.
+   * \param max_bin Maximum number of bins.
+   *
+   * \return A created quantile based DMatrix.
+   */
+  template <typename DataIterHandle, typename DMatrixHandle,
+            typename DataIterResetCallback, typename XGDMatrixCallbackNext>
+  static DMatrix *Create(DataIterHandle iter, DMatrixHandle proxy,
```
Comment (RAMitchell) [inline include/xgboost/data.h:525]: How would this interface be used to create a CPU quantile DMatrix?

File: src/data/iterative_device_dmatrix.cu [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    } else {
+      Dispatch(proxy, [&](auto const &value) {
+          common::AdapterDeviceSketch(value, batch_param_.max_bin, missing,
+                                      device, p_sketch);
+        });
+    }
+
+    auto batch_rows = num_rows();
+    accumulated_rows += batch_rows;
+    dh::caching_device_vector<size_t> row_counts(batch_rows + 1, 0);
```
Comment (RAMitchell) [inline src/data/iterative_device_dmatrix.cu:93]: I would recommend dh::TemporaryArray if the size is constant, it doesn't introduce synchronisation.

File: src/data/iterative_device_dmatrix.cu [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      summary_array.at(c).Reduce(summary, nbytes);
+    }
+  }
+  sketch_containers.clear();
+
+  // Build the final summary.
+  std::vector<common::WQSketch> sketches(cols);
+#pragma omp parallel for num_threads(nthread) if (nthread > 0)
+  for (omp_ulong c = 0; c < cols; ++c) {
+    sketches.at(c).Init(
```
Comment (RAMitchell) [inline src/data/iterative_device_dmatrix.cu:127]: So this is the 'old' sketching method, and your GK pr will change this?

File: include/xgboost/data.h (hunk 2 above)
Comment (trivialfis) [reply include/xgboost/data.h:525]: If we were to support CPU, the iterative DMatrix can dispatch the initialization based on source of data.

File: include/xgboost/c_api.h (hunk 1 above)
Comment (trivialfis) [reply include/xgboost/c_api.h:164]: Oops.

**Comment by trivialfis** [comment]
>  will device DMatrix be removed in future, or do you intend to keep both

Will be removed in the next extracted PR.

File: src/data/iterative_device_dmatrix.cu (hunk 4 above)
Comment (trivialfis) [reply src/data/iterative_device_dmatrix.cu:127]: Yup.  After the GK PR, whole pipeline is on GPU.

File: src/data/iterative_device_dmatrix.cu (hunk 3 above)
Comment (trivialfis) [reply src/data/iterative_device_dmatrix.cu:93]: It's not constant, each batch can have different rows.  Actually they don't even have to be the same type as you can pass 1 cudf df and 1 cupy array, we will just dispatch them to different adapter and run sketching accordingly.  Not really useful in practice, but we don't have any limitation on input data.

**Comment by trivialfis** [comment]
https://travis-ci.org/github/dmlc/xgboost/jobs/704501817

**Pull request merged**